miniapps/dpg/util/complexweakform.cpp

```diff
@@ -0,0 +1,1079 @@
+   for (int i = 0; i<mat_r->NumRowBlocks(); i++)
+   {
+      int h = dof_offsets[i+1] - dof_offsets[i];
+      for (int j = 0; j<mat_r->NumColBlocks(); j++)
+      {
+         int w = dof_offsets[j+1] - dof_offsets[j];
+         mat_r->SetBlock(i,j,new SparseMatrix(h, w));
+         mat_i->SetBlock(i,j,new SparseMatrix(h, w));
+      }
+   }
+   y = new Vector(2*dof_offsets.Last());
+   double * ydata = y->GetData();
+   *y=0.;
+   y_r = new BlockVector(ydata, dof_offsets);
+   y_i = new BlockVector(&ydata[dof_offsets.Last()], dof_offsets);
+}
+
+void ComplexDPGWeakForm::Finalize(int skip_zeros)
+{
+   if (mat_r)
+   {
+      mat_r->Finalize(skip_zeros);
+      mat_i->Finalize(skip_zeros);
+   }
+   if (mat_e_r)
+   {
+      mat_e_r->Finalize(skip_zeros);
+      mat_e_i->Finalize(skip_zeros);
+   }
+   if (static_cond) { static_cond->Finalize(); }
+}
+
+/// Adds new Domain BF Integrator. Assumes ownership of @a bfi.
+void ComplexDPGWeakForm::AddTrialIntegrator(
+   BilinearFormIntegrator *bfi_r,
+   BilinearFormIntegrator *bfi_i,
+   int n, int m)
+{
+   MFEM_VERIFY(n<trial_fes.Size(),
+               "ComplexDPGWeakFrom::AddTrialIntegrator: trial fespace index out of bounds");
+   MFEM_VERIFY(m<test_fecols.Size(),
+               "ComplexDPGWeakFrom::AddTrialIntegrator: test fecol index out of bounds");
+   if (bfi_r)
+   {
+      trial_integs_r(n,m)->Append(bfi_r);
+   }
+   if (bfi_i)
+   {
+      trial_integs_i(n,m)->Append(bfi_i);
+   }
+}
+
+/// Adds new Domain BF Integrator. Assumes ownership of @a bfi.
+void ComplexDPGWeakForm::AddTestIntegrator(
+   BilinearFormIntegrator *bfi_r,
+   BilinearFormIntegrator *bfi_i,
+   int n, int m)
+{
+   MFEM_VERIFY(n<test_fecols.Size() && m<test_fecols.Size(),
+               "ComplexDPGWeakFrom::AdTestIntegrator: test fecol index out of bounds");
+   if (bfi_r)
+   {
+      test_integs_r(n,m)->Append(bfi_r);
+   }
+   if (bfi_i)
+   {
+      test_integs_i(n,m)->Append(bfi_i);
+   }
+}
+
+/// Adds new Domain LF Integrator. Assumes ownership of @a bfi.
+void ComplexDPGWeakForm::AddDomainLFIntegrator(
+   LinearFormIntegrator *lfi_r,
+   LinearFormIntegrator *lfi_i, int n)
+{
+   MFEM_VERIFY(n<test_fecols.Size(),
+               "ComplexDPGWeakFrom::AddDomainLFIntegrator: test fecol index out of bounds");
+   if (lfi_r)
+   {
+      lfis_r[n]->Append(lfi_r);
+   }
+   if (lfi_i)
+   {
+      lfis_i[n]->Append(lfi_i);
+   }
+}
+
+void ComplexDPGWeakForm::BuildProlongation()
+{
+   P = new BlockMatrix(dof_offsets, tdof_offsets);
+   R = new BlockMatrix(tdof_offsets, dof_offsets);
+   P->owns_blocks = 0;
+   R->owns_blocks = 0;
+   for (int i = 0; i<nblocks; i++)
+   {
+      const SparseMatrix *P_ = trial_fes[i]->GetConformingProlongation();
+      if (P_)
+      {
+         const SparseMatrix *R_ = trial_fes[i]->GetRestrictionMatrix();
+         P->SetBlock(i,i,const_cast<SparseMatrix*>(P_));
+         R->SetBlock(i,i,const_cast<SparseMatrix*>(R_));
+      }
+      else
+      {
+         // do nothing
```

Comment:
Seems kinda weird to have an explicitr do nothing here insead of just dropping the else clause

Reply:
Indeed. Fixed